File: problems/0109_Convert_Sorted_List_to_Binary_Search_Tree/leetcode109.java
class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        ArrayList<Integer> list=new ArrayList<>();
        while(head!=null){
        	list.add(head.val);
        	head=head.next;
        }
        return build(list,0,list.size()-1);
    }
    private TreeNode build(ArrayList<Integer> nums,int head,int tail){
		if(head<tail){
			int index=(head+tail+1)/2;
			TreeNode root=new TreeNode(nums.get(index));
			root.left=build(nums,head,index-1);
			root.right=build(nums,index+1,tail);
			return root;
		}
		else if(head>tail){
			return null;
		}else{
			return new TreeNode(nums.get(head));
		}	
	}

}

File: problems/0109_Convert_Sorted_List_to_Binary_Search_Tree/leetcode109.py
# !/usr/bin/env python3
# -*- coding:utf-8 -*-

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedListToBST(self, head):  # 168ms
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        nums = []
        while head:
            nums.append(head.val)
            head = head.next
        return self.hhhhh(nums)

    def hhhhh(self, nums):
        if not nums:
            return None
        _len = len(nums)
        mid = _len // 2
        root = TreeNode(nums[mid])
        root.left = self.hhhhh(nums[:mid])
        root.right = self.hhhhh(nums[mid + 1:])
        return root